src/psij/_plugins.py

```diff
@@ -0,0 +1,120 @@
+import importlib
+import logging
+from bisect import bisect_left
+from distutils.versionpredicate import VersionPredicate
+from types import ModuleType
+from typing import Tuple, Dict, List, Union, Type, Any, Optional, TypeVar
+
+from psij._descriptor import _Descriptor, _VersionEntry
+
+logger = logging.getLogger(__name__)
+
+
+def _getfile(o: Optional[Union[Type[Any], ModuleType]]) -> str:
+    if o is None:
+        return ''
+    else:
+        return o.__psij_file__  # type: ignore
+
+
+def _split_cls_name(cls: str) -> Tuple[str, str]:
+    comps = cls.split('.')
+    return '.'.join(comps[:-1]), comps[-1]
+
+
+T = TypeVar('T')
+
+
+def _register_plugin(desc: _Descriptor, root_path: str, type: str,
+                     store: Dict[str, List[_VersionEntry[T]]]) -> None:
+    """
+    Registers a class with a certain base class through a :class:`~psij._Descriptor`.
+
+    This is used internally to dynamically find and register :class:`~psij.JobExecutor` and
+    :class:`~psij.Launcher` classes.
+
+    Parameters
+    ----------
+    desc:
+        a descriptor whose `cls` property must point to a class that has `cls` as a base class.
+    type:
+        the type of class being registered (either "executor" or "launcher")
+    store:
+        a dictionary where the registered plugins are being stored
+    """
+    module, cls_name = _split_cls_name(desc.cls)
+    cls = None
+    exc = None
+    mod_path = root_path + '/' + module.replace('.', '/') + '.py'
+    try:
+        spec = importlib.util.spec_from_file_location(module, mod_path)
+        if spec is None:
+            raise Exception('Failed to load %s' % mod_path)
+        mod = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(mod)  # type: ignore
+        cls = getattr(mod, cls_name)
+        cls.__psij_file__ = mod_path
+    except Exception as ex:
+        s = str(ex)
+        logger.info(s)
+        exc = ex
+
+    if desc.name not in store:
+        store[desc.name] = []
+    existing = store[desc.name]
+    entry = _VersionEntry(desc.version, desc_path=desc.path, plugin_path=mod_path,
+                          ecls=cls, exc=exc)  # type: _VersionEntry[T]
+    # check if an object with this version already exists
```

Comment:
so does this code allowed registration of multiple versions of the same-named executor?

Reply:
Yes.